Add sound mode support

starkillerOG opened the pull request:
## Description:
Add sound mode support for the DenonAVR platform.
Using the just merged [general sound mode support](https://github.com/home-assistant/home-assistant/commit/f6963315631bdecd7f6acca41369eeb0b9b473bb).

Added backend support for selecting the sound mode and updating the sound mode status.

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#5521

## Example entry for `configuration.yaml` (if applicable):
```yaml
media_player:
  - platform: denonavr
    host: IP_ADRESS
    name: "Marantz"
    show_all_sources: False
    timeout: 2
    sound_mode: True
    sound_mode_dict: {'MUSIC':['PLII MUSIC'], 'MOVIE':['PLII MOVIE'], 'GAME':['PLII GAME'], 'PURE DIRECT':['DIRECT'], 'AUTO':['None'], 'DOLBY DIGITAL':['DOLBY DIGITAL'], 'MCH STEREO':['MULTI CH STEREO'], 'STEREO':['STEREO']}
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [x] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

starkillerOG committed:
Add sound mode support

starkillerOG committed:
continuation line indent

starkillerOG committed:
indentation

starkillerOG committed:
indentation

starkillerOG commented:
@balloob Can you merge this PR for the denonAVR implementation of sound mode support?
(this is a follow up of the [general sound mode support](https://github.com/home-assistant/home-assistant/pull/14729) that you merged last week)

balloob commented on homeassistant/components/media_player/denonavr.py:
```diff
@@ -147,6 +160,24 @@ def __init__(self, receiver):
+        if sound_mode_support:
+            self._sound_mode = self._receiver.sound_mode
+            self._sound_mode_raw = self._receiver.sound_mode_raw
+            if sound_mode_dict is None:
```
If we can fetch it from the device, we should not allow it to be configured. Integrations should only represent the raw information that we receive from the device.

starkillerOG replied:
@balloob the sound_mode_dict can not be fetched from the device (receiver) itself, the denonavr library contains an default sound_mode_dict that is hard coded in the denonavr library. However that dictionary is based upon the data of about 10 diffrent models of denon receivers (from which the information xml files are available in the denonavr git repositry, but their could potentially be other values available for newer or very old receivers.
This allows for flexibility for these older or newer receivers.

balloob replied:
I think that we should not include this config option because it will not motivate people to update the dictionary upstream for old and new receivers, resulting in all users having to collect the same information.

starkillerOG committed:
Remove option to configure sound_mode_dict

starkillerOG commented:
@balloob I removed the option to configure the sound_mode_dict.
Do you have any aditional improvements/feedback?

starkillerOG commented:
@balloob can this now be merged?

MartinHjelmare commented on homeassistant/components/media_player/denonavr.py:
```diff
@@ -197,12 +223,27 @@ def source_list(self):
+        return self._sound_mode
+
+    @property
+    def sound_mode_raw(self):
```
This isn't used.

starkillerOG committed:
Sound mode support

- removed the sound_mode_raw propertie because it was not used, (still available through self._sound_mode_raw (as device attribute for automations and diagnostics)

starkillerOG commented:
@MartinHjelmare Just removed the un-used sound_mode_raw propertie.
Thanks for the review!
Any adittional feedback?

balloob commented on homeassistant/components/media_player/denonavr.py:
```diff
@@ -49,6 +53,7 @@
 PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
     vol.Optional(CONF_HOST): cv.string,
     vol.Optional(CONF_NAME): cv.string,
+    vol.Optional(CONF_SOUND_MODE, default=DEFAULT_SOUND_MODE): cv.boolean,
```
Why is sound mode configurable? Can it not be detected from the device?

starkillerOG replied:
This is the option to disable sound_mode_support for old types receivers that do not have sound modes. As far as I know this can not be detected from the receiver itself. At least their is no direct bolean if sound mode is supported. It might be possible to see it from the absense of the current sound mode if you request it from the receiver.

However I always like to have the option to disable components in case they give problems for certain types/models. Of course I only have my own receiver to test with and that one does support sound_mode. So I cann't really test with a receiver that doesn't support sound mode.

balloob replied:
> However I always like to have the option to disable components in case they give problems for certain types/models. 

In Home Assistant we aim to represent devices as-is. If we go down the route to disable sound modes, we can start adding config options for everything…

Please remove the config option and base it on detecting support.

balloob replied:
Okay. This PR can be merged when the config option has been removed and the lib has been updated.

starkillerOG commented:
@balloob can this denonavr sound mode support be merged?

fabaff commented:
The library update and the removal of the option is not done as mentioned [here](https://github.com/home-assistant/home-assistant/pull/14910/files#r195424942).

starkillerOG committed:
Detect sound mode support from device

Removed the config option to indicate if sound mode is supported.
Added detection if sound mode is supported from the receiver itself.
Pushed denonavr library to V.0.7.4

starkillerOG committed:
Pushed denonavr to v.0.7.4

starkillerOG commented:
@balloob Thank you for all your help througout this PR!

Martinvdm commented:
After update to 0.74 the log is beeing spammed with the following items:
```
2018-07-20 21:51:12 WARNING (SyncWorker_9) [DenonAVR] Not able to match sound mode, returning raw sound mode.
2018-07-20 21:51:23 WARNING (SyncWorker_13) [DenonAVR] Not able to match sound mode, returning raw sound mode.
2018-07-20 21:51:34 WARNING (SyncWorker_5) [DenonAVR] Not able to match sound mode, returning raw sound mode.
2018-07-20 21:51:45 WARNING (SyncWorker_7) [DenonAVR] Not able to match sound mode, returning raw sound mode.
2018-07-20 21:51:56 WARNING (SyncWorker_17) [DenonAVR] Not able to match sound mode, returning raw sound mode.
2018-07-20 21:52:07 WARNING (SyncWorker_3) [DenonAVR] Not able to match sound mode, returning raw sound mode.
2018-07-20 21:52:18 WARNING (SyncWorker_17) [DenonAVR] Not able to match sound mode, returning raw sound mode.
2018-07-20 21:52:29 WARNING (SyncWorker_14) [DenonAVR] Not able to match sound mode, returning raw sound mode.
2018-07-20 21:52:40 WARNING (SyncWorker_12) [DenonAVR] Not able to match sound mode, returning raw sound mode.
```

MartinHjelmare commented:
Please open an issue.
